lib/events/stream.go

```diff
@@ -650,20 +690,39 @@ func (w *sliceWriter) newSlice() (*slice, error) {
 	// This buffer will be returned to the pool by slice.Close
 	buffer := w.proto.cfg.BufferPool.Get()
 	buffer.Reset()
+
+	var encrypted bool
+	var writer io.WriteCloser = &bufferCloser{Buffer: buffer}
+	if w.encrypter != nil {
+		// we want to encrypt after compression, so gzip needs to be the outermost layer
+		encryptedWriter, err := w.encrypter.WithEncryption(w.proto.completeCtx, writer)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		encrypted = writer != encryptedWriter
+		writer = encryptedWriter
```

Comment:
Re: returning an error in `WithEncryption` instead of returning the plain writer:

If `encrypter` is set, does that mean encryption should be enforced? Shouldn't we return an error here when `writer == encryptedWriter` rather than just continuing without encryption?

Assuming we do want to proceed without encryption, the logic here is a bit awkward to follow. I would rather see an error returned from `WithEncryption` saying encryption is not configured and an error check here to set `encrypted=false` and continue with the plain writer.